src/Simplex/Messaging/Agent/Client.hs

```diff
@@ -152,12 +177,26 @@ subscribeQueue c rq@ReceiveQueue {server, rcvPrivateKey, rcvId} connAlias = do
   addSubscription c rq connAlias
 
 addSubscription :: MonadUnliftIO m => AgentClient -> ReceiveQueue -> ConnAlias -> m ()
-addSubscription c ReceiveQueue {server, rcvId} connAlias =
-  atomically . modifyTVar (subscribed c) $ M.insert connAlias (server, rcvId)
+addSubscription c ReceiveQueue {server} connAlias = atomically $ do
+  modifyTVar (subscrConns c) $ M.insert connAlias server
+  modifyTVar (subscrSrvrs c) $ M.alter (Just . addSub) server
+  where
+    addSub :: Maybe (Set ConnAlias) -> Set ConnAlias
+    addSub (Just cs) = S.insert connAlias cs
+    addSub _ = S.singleton connAlias
 
 removeSubscription :: AgentMonad m => AgentClient -> ConnAlias -> m ()
-removeSubscription c connAlias =
-  atomically . modifyTVar (subscribed c) $ M.delete connAlias
+removeSubscription AgentClient {subscrConns, subscrSrvrs} connAlias = atomically $ do
+  cs <- readTVar subscrConns
+  writeTVar subscrConns $ M.delete connAlias cs
```

Comment:
Is it possible to reuse some functionality between `clientDisconnected` and `removeSubscription`?

Also what is the reason that it is `writeTVar subscrConns $ M.delete connAlias cs` in `removeSubscription`, but `modifyTVar (subscrConns c) $ M.delete connAlias` in `removeSubscription`? Maybe add small comments explaining the intention?

Reply:
>Is it possible to reuse some functionality between clientDisconnected and removeSubscription?

maybe, not clear... any suggestion?

> Also what is the reason that it is writeTVar subscrConns $ M.delete connAlias cs in removeSubscription, but modifyTVar (subscrConns c) $ M.delete connAlias in removeSubscription? Maybe add small comments explaining the intention?

I use this cs in the first case to remove subscriptions from the servers, the second only modifies it.